### PR TITLE
View container layout changes

### LINKED ISSUE #5175: [debug][ui/ux] Align the Debug View with VS Code
I have created this as a subtask of #5058.

### PR DESCRIPTION
**Pull request by spoenemann**
Fixes #5175. This is a continuation of #5392.

### CODE REVIEW

**Comment by kittaakos** [comment]
I am trying this now. Please note the linter error; the CIs are red.

**Comment by kittaakos** [comment]
`Hide` > `Show` does not respect the previous state, but always expands the view container part:
screencast 2019-06-26 10-34-12

File: packages/core/src/browser/view-container.ts [hunk 1]
```diff
@@ -0,0 +1,935 @@
...
+        export interface WidgetOptions {
+            readonly order?: number;
+            readonly weight?: number;
+            readonly initiallyCollapsed?: boolean;
+            readonly canHide?: boolean;
+            readonly initiallyHidden?: boolean;
+        }
+
+        export interface WidgetDescriptor {
+            // tslint:disable-next-line:no-any
```
Comment (kittaakos) [inline packages/core/src/browser/view-container.ts]: This seems to be obsolete.

File: packages/core/src/browser/view-container.ts [hunk 2]
```diff
@@ -0,0 +1,935 @@
...
+            readonly options?: WidgetOptions;
+        }
+
+    }
+
+    export interface State {
+        parts: ViewContainerPart.State[]
+    }
+
+    export function getOrientation(node: HTMLElement) {
```
Comment (kittaakos) [inline packages/core/src/browser/view-container.ts]: Please add the explicit return type.

File: packages/core/src/browser/view-container.ts [hunk 3]
```diff
@@ -0,0 +1,935 @@
...
+        header.appendChild(title);
+
+        if (ViewContainerPart.ContainedWidget.is(this.wrapped)) {
+            for (const { tooltip, execute, className } of this.wrapped.toolbarElements.filter(e => e.enabled !== false)) {
+                const toolbarItem = document.createElement('span');
+                toolbarItem.classList.add('element');
+                if (typeof className === 'string') {
+                    toolbarItem.classList.add(...className.split(' '));
+                } else if (Array.isArray(className)) {
+                    toolbarItem.classList.add(...className);
```
Comment (kittaakos) [inline packages/core/src/browser/view-container.ts]: This is unsafe; the array of CSS classes can be `['this is incorrect', 'correct']`.

File: packages/core/src/browser/view-container.ts [hunk 4]
```diff
@@ -0,0 +1,935 @@
...
+        }
+        super.onUpdateRequest(msg);
+    }
+
+}
+
+export namespace ViewContainerPart {
+
+    /**
+     * Make sure to adjust the `line-height` of the `.theia-view-container .part .header` CSS class when modifying this, and vice versa.
```
Comment (kittaakos) [inline packages/core/src/browser/view-container.ts]: Please align the doc with the CSS: https://github.com/theia-ide/theia/pull/5536/commits/8dea8fdea59e634eea9dc60e50dd433021c00d9e#diff-a83c42ebfd482eb91ce1279d274c64b1R4 or delete it. Thanks!

File: packages/core/src/browser/view-container.ts [hunk 5]
```diff
@@ -0,0 +1,935 @@
...
+                style.maxHeight = null;
+            }
+        }
+        super.onFitRequest(msg);
+    }
+
+    /**
+     * Sinusoidal tween function for smooth animation.
+     */
+    protected tween(t: number) {
```
Comment (kittaakos) [inline packages/core/src/browser/view-container.ts]: Missing return type.

**Comment by spoenemann** [comment]
> Hide > Show does not respect the previous state, but always expands the view container part

I did that intentionally. I thought when I "unhide" a part I want to see its content, too. Do you think we should change that?

**Comment by spoenemann** [comment]
> the collapse/expand behavior is slightly different from the VS Code one.

Yes, that's due to the default _fit_ behavior of SplitLayout, which distributes the gained space across all expanded parts. I think it's not bad.

**Comment by kittaakos** [comment]
> Do you think we should change that?

If it was on purpose; no.

**Comment by akosyakov** [comment]
@spoenemann Could you address @kittaakos comments and build failures please?

**Comment by spoenemann** [comment]
Done.

**Comment by akosyakov** [comment]
Something wrong with the first section of the debug view, there are scrolls:
Screen Shot 2019-07-05 at 10 33 25
Screen Shot 2019-07-05 at 10 33 23

Rather than that it is super good change. Cannot wait to integrate it in the plugin system.

**Comment by akosyakov** [comment]
We should review how we contribute toolbar items. We should do it via menu contributions, i.e. there should be new menu path for each view container part which is displayed as a toolbar. It would match to VS Code better and does not require new interfaces as now. But it can be done as a separate PR.

**Comment by spoenemann** [comment]
I don't have time to finish this before my vacation.

**Comment by kittaakos** [comment]
> I don't have time to finish this before my vacation.

I'll take care of it later today.

**Comment by kittaakos** [comment]
> I'll take care of it later today.

It's done.

**Comment by akosyakov** [comment]
Actually i see scrolls on master as well, please merge.

**Pull request merged**